### PR TITLE
OpenFL 4 compatibility

### PR DESCRIPTION
**Pull request by Beeblerox**
it works only with openfl 4.5.0.
i can't get it to work with higher versions - compiler just fails silently for some reason

### CODE REVIEW

File: haxelib.json [hunk 1]
```diff
@@ -4,8 +4,8 @@
 	"license": "MIT",
 	"tags": ["game", "openfl", "flash", "html5", "neko", "cpp", "android", "ios", "cross"],
 	"description": "HaxeFlixel is a 2D game engine based on OpenFL that delivers cross-platform games.",
-	"version": "4.2.0",
```
Comment (Gama11) [inline haxelib.json]: I guess version and releasenote were changed accidentally?

Comment (Beeblerox) [inline haxelib.json]: @Gama11 yeah, sorry. will revert this change

File: flixel/graphics/FlxRenderTexture.hx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package flixel.graphics;
+
+import flixel.util.FlxDestroyUtil.IFlxDestroyable;
+
+// TODO: implement it... :)
```
Comment (Gama11) [inline flixel/graphics/FlxRenderTexture.hx]: What about this?

File: flixel/graphics/shaders/tiles/FlxColored.hx [hunk 3]
```diff
@@ -0,0 +1,44 @@
+package flixel.graphics.shaders.tiles;
+
+import flixel.graphics.shaders.FlxBaseShader;
+
+/**
+ * Default shader used by batcher for rendering quads without textures.
+ */
+class FlxColored extends FlxBaseShader
```
Comment (Gama11) [inline flixel/graphics/shaders/tiles/FlxColored.hx]: This seems like a bit of a weird class name. Wouldn't `FlxColoredShader` or `FlxColorShader` be more fitting?

File: flixel/graphics/shaders/triangles/FlxColored.hx [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package flixel.graphics.shaders.triangles;
+
+import flixel.graphics.shaders.FlxBaseShader;
+
+/**
+ * Default shader used for rendering triangles with colored vertices (each vertice could have inidividual color)
+ */
+class FlxColored extends FlxBaseShader
```
Comment (Gama11) [inline flixel/graphics/shaders/triangles/FlxColored.hx]: Same here.

File: flixel/graphics/FlxRenderTexture.hx (hunk 2 above)
Comment (Beeblerox) [inline flixel/graphics/FlxRenderTexture.hx]: oops, will remove this.
it's from plans for future features

File: flixel/graphics/shaders/triangles/FlxTextured.hx [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package flixel.graphics.shaders.triangles;
+
+import flixel.graphics.shaders.FlxBaseShader;
+
+/**
+ * Default shader used for rendering textured triangles with applied color transform of the FlxStrip.
+ */
+class FlxTextured extends FlxBaseShader
```
Comment (Gama11) [inline flixel/graphics/shaders/triangles/FlxTextured.hx]: And here.

File: flixel/system/FlxPreloader.hx [hunk 6]
```diff
@@ -9,10 +9,10 @@ import flash.text.TextField;
 import flash.text.TextFormat;
 import flixel.FlxG;
 
-@:keep @:bitmap("assets/images/preloader/light.png")
```
Comment (Gama11) [inline flixel/system/FlxPreloader.hx]: This seems like an unintentional change? This undoes a DCE fix of mine.

File: flixel/system/FlxPreloader.hx [hunk 7]
```diff
@@ -48,20 +48,18 @@ class FlxPreloader extends FlxBasePreloader
...
 	{
+		#if !js
 		_buffer = new Sprite();
 		_buffer.scaleX = _buffer.scaleY = 2;
 		addChild(_buffer);
 		_width = Std.int(Lib.current.stage.stageWidth / _buffer.scaleX);
 		_height = Std.int(Lib.current.stage.stageHeight / _buffer.scaleY);
 		_buffer.addChild(new Bitmap(new BitmapData(_width, _height, false, 0x00345e)));
-		
-		var logoLight = createBitmap(GraphicLogoLight, function(logoLight:Bitmap)
```
Comment (Gama11) [inline flixel/system/FlxPreloader.hx]: And this was a fix for HTML5 IIRC.

File: flixel/system/FlxPreloader.hx (hunk 6 above)
Comment (Beeblerox) [inline flixel/system/FlxPreloader.hx]: yes, sorry about this

File: flixel/FlxGame.hx [hunk 8]
```diff
@@ -881,4 +893,33 @@ class FlxGame extends Sprite
 		// expensive, only call if necessary
 		return Lib.getTimer();
 	}
-}
+	
+	#if FLX_RENDER_GL
+	private var renderHelper(get, null):GLRenderHelper;
```
Comment (Gama11) [inline flixel/FlxGame.hx]: Shouldn't this be `get, never`?

File: flixel/FlxSprite.hx [hunk 9]
```diff
@@ -630,6 +631,9 @@ class FlxSprite extends FlxObject
 	{
 		if (_frame == null)
 			loadGraphic("flixel/images/logo/default.png");
+		
+	//	if (_frame == null)
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: ?

File: flixel/system/macros/FlxDefines.hx [hunk 10]
```diff
@@ -90,12 +90,12 @@ class FlxDefines
 		abortMinVersion("Lime", "2.8.1", (macro null).pos);
 		#end
 
-		#if (openfl >= "4.0.0")
-		abortMaxVersion("OpenFL", "4.0.0", "3.6.1", (macro null).pos);
+		#if (openfl > "4.5.0")
+		abortMaxVersion("OpenFL", "4.5.1", "4.5.0", (macro null).pos);
```
Comment (Gama11) [inline flixel/system/macros/FlxDefines.hx]: Hm, we definitely need to figure out what the issue with the latest versions is.

File: flixel/system/render/blit/FlxBlitView.hx [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+	
+	override private function set_angle(Angle:Float):Float 
+	{
+		if (flashSprite != null)
+			flashSprite.rotation = Angle;
+		
+		return Angle;
+	}
+	
+	override public function set_smoothing(Smoothing:Bool):Bool
```
Comment (Gama11) [inline flixel/system/render/blit/FlxBlitView.hx]: Setter, should probably be private?

Comment (Beeblerox) [inline flixel/system/render/blit/FlxBlitView.hx]: of course!
But it's strange, because this method is private in super class. I didn't know that you could change method visibility modifier in subclasses

Comment (Gama11) [inline flixel/system/render/blit/FlxBlitView.hx]: You can, but only from `private` to `public`. The other direction would cause issues.

File: flixel/system/macros/FlxDefines.hx (hunk 10 above)
Comment (Beeblerox) [inline flixel/system/macros/FlxDefines.hx]: agree, but had no luck to find out where is the problem

File: flixel/system/render/common/FlxCameraView.hx [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+	/**
+	 * Whether to batch drawTriangles() calls or not.
+	 * Default value is true.
+	 */
+	public static var BATCH_TRIANGLES:Bool = #if FLX_RENDER_GL false #else true #end;
+	
+	/**
+	 * Tracks total number of `drawTiles()` calls made each frame.
+	 */
+	public static var _DRAWCALLS:Int = 0;
```
Comment (Gama11) [inline flixel/system/render/common/FlxCameraView.hx]: I guess this was always called `_DRAWCALLS`? Weird naming, normally we have a leading underscore for private vars, and it's not constant either, so it probably shouldn't be upper-cased...

Maybe a good opportunity to rename it to `drawCalls`?

File: flixel/system/render/common/FlxDrawBaseCommand.hx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+			&& this.graphics == graphic 
+			&& this.colored == colored
+			&& this.hasColorOffsets == hasColorOffsets
+			&& this.blending == blend
+			&& this.smoothing == smooth
+			&& this.repeat == repeat
+			&& this.shader == shader);
+	}
+	
+	public /*inline*/ function set(graphic:FlxGraphic, colored:Bool, hasColorOffsets:Bool = false,
```
Comment (Gama11) [inline flixel/system/render/common/FlxDrawBaseCommand.hx]: Commented-out `inline`?

Comment (Beeblerox) [inline flixel/system/render/common/FlxDrawBaseCommand.hx]: will remove it.
i'm overriding this method in one of the sublasses

File: flixel/system/render/common/FlxCameraView.hx (hunk 12 above)
Comment (Beeblerox) [inline flixel/system/render/common/FlxCameraView.hx]: ok, will do

**Comment by starry-abyss** [comment]
So is this going to be merged soon? Because when I tried the previous PR recently, it was terribly slow for From the Deep (neko), and had problems compiling on some other projects.

**Comment by Gama11** [comment]
When I run Mode on Neko, it crashes after clicking the "Play"-button:

```
Called from ? line 1
Called from ApplicationMain.hx line 78
Called from ApplicationMain.hx line 142
Called from lime/app/Application.hx line 212
Called from lime/_backend/native/NativeApplication.hx line 139
Called from lime/_backend/native/NativeApplication.hx line 376
Called from lime/app/Event.hx line 150
Called from openfl/display/Stage.hx line 207
Called from openfl/display/Stage.hx line 1028
Called from openfl/display/Stage.hx line 1188
Called from C:\HaxeToolkit\haxe\std/neko/Lib.hx line 65
Called from openfl/display/Stage.hx line 977
Called from openfl/display/Stage.hx line 1050
Called from openfl/display/DisplayObject.hx line 328
Called from openfl/events/EventDispatcher.hx line 213
Called from flixel/FlxGame.hx line 562
Called from flixel/FlxGame.hx line 862
Called from flixel/system/render/hardware/FlxHardwareView.hx line 262
Called from flixel/system/render/hardware/FlxHardwareView.hx line 240
Called from flixel/system/render/common/FlxDrawStack.hx line 255
Called from flixel/system/render/hardware/gl/FlxDrawQuadsCommand.hx line 223
Uncaught exception - Invalid field access : byteOffset)
```

Also, the camera shake effect is missing from the title screen.

OpenFL 4.5.0 and Lime 3.5.0, Haxe 3.4.0-rc.2.

**Comment by Gama11** [comment]
>i can't get it to work with higher versions - compiler just fails silently for some reason

Could you describe that in more detail? Have you tried compiling with `-verbose`?

For, Mode runs fine on Neko with OpenFL 4.6.0 and Lime 3.6.2. Well, apart from the crash I described.

What Haxe version are you using?

**Comment by Beeblerox** [comment]
omg. Mode on neko is soo slow. 
and i've spend so much time on this. total disgrace

**Comment by starry-abyss** [comment]
At least it didn't get merged!
And also Flash target support looks very good :-)

**Comment by Beeblerox** [comment]
@Gama11 project compiled after upgrading to OpenFL 4.6.0 and Lime 3.6.2 and rebooting machine
don't know why it didn't work with previous versions.

Now i need to optimize this trash

**Comment by Beeblerox** [comment]
@Gama11 Mode demo crashes after i press the Play button. but i don't get details, about what caused the crash (like in your case it's something with Float32Array manipulation).
There is almost no output, just the line:
```
AL lib: (EE) alc_cleanup: 1 device not closed
```
Will try to test it on another machine

**Comment by larsiusprime** [comment]
Don't be discouraged, we had the same issue with next, and it might even be openfl's fault 😜. You're doing great work!

**Comment by starry-abyss** [comment]
@Beeblerox Neko not showing the call stack on crash has something to do with HaxeDevelop, try 'lime test neko' manually, should work

**Comment by Beeblerox** [comment]
@starry-abyss thanks for advice, it helped!

@larsiusprime thank you for your kind words, i will try to improve perfomance

**Comment by MSGhero** [comment]
Is this going to be part of 4.3 or 5.0? i.e. will this be merged as soon as possible or put on hold to put as many breaking changes in 5.0 release as possible?

**Comment by starry-abyss** [comment]
I noticed that this game (https://github.com/starry-abyss/ld34) is like the only my project that both compiles (with this PR) and has performance issues on neko. It also hangs for several seconds before starting each state.
And the clue is two FlxTilemap's. If I don't add them to the state, the performance is good (1.5ms for draw in flixel debugger).
But I also have tilemaps in other projects performing good, so can't tell the difference.

File: flixel/FlxSprite.hx [hunk 14]
```diff
@@ -56,6 +57,7 @@ class FlxSprite extends FlxObject
 	 */
 	public var useFramePixels(default, set):Bool = true;
 	
+	// TODO: rename this to smoothing...
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: Hm? Isn't that a bit of an unnecessary breaking change?

Comment (Beeblerox) [inline flixel/FlxSprite.hx]: maybe, but this var responsible for smoothing, not for aliasing

Comment (Gama11) [inline flixel/FlxSprite.hx]: Hasn't it been like that since Flixel 2 (AS3) at least? Could always add a note about that to the docs.

If we do decide to change it, we should probably keep `antialiasing` as an alias for `smoothing` and mark it as `@:deprecated` to remain backwards-compatible.

Comment (Beeblerox) [inline flixel/FlxSprite.hx]: agree, let's mark it as deprecated

File: flixel/FlxSprite.hx [hunk 15]
```diff
@@ -57,11 +57,17 @@ class FlxSprite extends FlxObject
...
 	/**
 	 * Controls whether the object is smoothed when rotated, affects performance.
 	 */
-	public var antialiasing(default, set):Bool = false;
+	public var smoothing(default, set):Bool = false;
+	
+	/**
+	 * Controls whether the object is smoothed when rotated, affects performance.
+	 */
+	@:deprecated
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: We should probably put a deprecation message here ("Use smoothing instead" or something like that.)

File: flixel/graphics/FlxTrianglesData.hx [hunk 16]
```diff
@@ -14,8 +14,8 @@ import openfl.utils.Float32Array;
 import flixel.system.render.hardware.gl.GLUtils;
 #end
 
-// TODO: rename to FlxTrianglesData...
-class TrianglesData implements IFlxDestroyable
+// TODO: rename to FlxFlxTrianglesData...
```
Comment (Gama11) [inline flixel/graphics/FlxTrianglesData.hx]: :D

File: flixel/FlxSprite.hx (hunk 15 above)
Comment (Beeblerox) [inline flixel/FlxSprite.hx]: you're right

File: flixel/FlxSprite.hx [hunk 17]
```diff
@@ -59,11 +59,13 @@ class FlxSprite extends FlxObject
 	
 	/**
 	 * Controls whether the object is smoothed when rotated, affects performance.
+	 * @since 4.3.0
 	 */
 	public var smoothing(default, set):Bool = false;
 	
 	/**
 	 * Controls whether the object is smoothed when rotated, affects performance.
+	 * It's better to use `smoothing` property instead, since `antialiasing` is deprecated.
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: I meant as an argument for `@:deprecated`, then it will show up in compiler warnings. See here for an example: https://github.com/HaxeFlixel/flixel/blob/dev/flixel/math/FlxRandom.hx#L307

File: flixel/FlxSprite.hx [hunk 18]
```diff
@@ -59,11 +59,13 @@ class FlxSprite extends FlxObject
 	
 	/**
 	 * Controls whether the object is smoothed when rotated, affects performance.
+	 * @since 4.3.0
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: Nice! 👍

File: flixel/FlxSprite.hx [hunk 19]
```diff
@@ -65,9 +65,8 @@ class FlxSprite extends FlxObject
 	
 	/**
 	 * Controls whether the object is smoothed when rotated, affects performance.
-	 * It's better to use `smoothing` property instead, since `antialiasing` is deprecated.
 	 */
-	@:deprecated
+	@:deprecated("It's better to use `smoothing` property instead.")
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: Sorry to be so nit-picky, but "It's better to" to me implies that they somehow do different things, when it's really just been renamed.

File: flixel/FlxStrip.hx [hunk 20]
```diff
@@ -15,45 +21,158 @@ import flixel.graphics.tile.FlxDrawTrianglesItem.DrawData;
...
+		data = FlxDestroyUtil.destroy(data);
+		bounds = FlxDestroyUtil.put(bounds);
 		
 		super.destroy();
 	}
 	
 	override public function draw():Void 
 	{
-		if (alpha == 0 || graphic == null || vertices == null)
+		if (alpha == 0 || /*graphic == null ||*/ vertices == null)
```
Comment (gamedevsam) [inline flixel/FlxStrip.hx]: Why not just remove this commented code?

File: flixel/FlxSprite.hx (hunk 9 above)
Comment (gamedevsam) [inline flixel/FlxSprite.hx]: Better to remove than comment out, unless there is good reason to believe it will be useful in the future.

File: flixel/graphics/shaders/tiles/FlxColored.hx (hunk 3 above)
Comment (gamedevsam) [inline flixel/graphics/shaders/tiles/FlxColored.hx]: Perhaps ```FlxDefaultShader```?

File: flixel/graphics/shaders/tiles/FlxTextured.hx [hunk 21]
```diff
@@ -0,0 +1,61 @@
+package flixel.graphics.shaders.tiles;
+
+import flixel.graphics.shaders.FlxBaseShader;
+
+/**
+ * Default shader used by batcher for rendering textured quads.
+ */
+class FlxTextured extends FlxBaseShader
```
Comment (gamedevsam) [inline flixel/graphics/shaders/tiles/FlxTextured.hx]: I agree with Gama that more descriptive names would be better. For me the most important thing is consistency, so if the above class was to be renamed ```FlxColorShader``` than this one should be called ```FlxTextureShader```. You could also do ```FlxColoredShader``` and then this file would be called ```FlxTexturedShader```. But shorter names are also usually a better choice :-P

File: flixel/graphics/shaders/triangles/FlxSingleColored.hx [hunk 22]
```diff
@@ -0,0 +1,48 @@
+package flixel.graphics.shaders.triangles;
+
+import flixel.graphics.shaders.FlxBaseShader;
+
+/**
+ * Default shader for rendering triangles without textures and without vertex colors
+ * (each vertex color is defined only by color transform of the FlxStrip).
+ */
+class FlxSingleColored extends FlxBaseShader
```
Comment (gamedevsam) [inline flixel/graphics/shaders/triangles/FlxSingleColored.hx]: ```FlxSingleColor``` perhaps?

File: flixel/system/macros/FlxDefines.hx (hunk 10 above)
Comment (gamedevsam) [inline flixel/system/macros/FlxDefines.hx]: It's April, no luck still?

File: flixel/system/FlxPreloader.hx (hunk 7 above)
Comment (gamedevsam) [inline flixel/system/FlxPreloader.hx]: Perhaps not all recent changes have been merged with this? Quite an odd regression.

File: flixel/system/FlxPreloader.hx (hunk 6 above)
Comment (gamedevsam) [inline flixel/system/FlxPreloader.hx]: Revert!

File: flixel/graphics/FlxMaterial.hx [hunk 23]
```diff
@@ -0,0 +1,279 @@
+package flixel.graphics;
+
+/*
```
Comment (gamedevsam) [inline flixel/graphics/FlxMaterial.hx]: Added a commented out file, what's the reasoning?

File: flixel/system/render/hardware/FlxHardwareView.hx [hunk 24]
```diff
@@ -0,0 +1,327 @@
...
+		drawStack.fillRect(_fillRect, Color, FxAlpha);
+	}
+	
+	override public function drawFX(FxColor:FlxColor, FxAlpha:Float = 1.0):Void 
+	{
+		var alphaComponent:Float = FxColor.alpha;
+		fill(FxColor.to24Bit(), true, ((alphaComponent <= 0) ? 0xff : alphaComponent) * FxAlpha / 255);
+	}
+	
+	override public function lock(useBufferLocking:Bool):Void
```
Comment (gamedevsam) [inline flixel/system/render/hardware/FlxHardwareView.hx]: What is ```useBufferLocking``` for? It doesn't seem to get used anywhere.

File: flixel/system/render/hardware/FlxHardwareView.hx [hunk 25]
```diff
@@ -0,0 +1,327 @@
...
+		#end
+		
+		// Clearing camera's debug sprite
+		#if FLX_DEBUG
+		debugLayer.graphics.clear();
+		#end
+		fill(camera.bgColor.to24Bit(), camera.useBgAlphaBlending, camera.bgColor.alphaFloat);
+	}
+	
+	override public function unlock(useBufferLocking:Bool):Void
```
Comment (gamedevsam) [inline flixel/system/render/hardware/FlxHardwareView.hx]: Unused parameter ```useBufferLocking```.

File: flixel/system/render/hardware/gl/FlxDrawTrianglesCommand.hx [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+#else
+class FlxDrawTrianglesCommand extends FlxDrawHardwareCommand<FlxDrawTrianglesCommand>
+{
+	public var data:FlxTrianglesData;
+	public var matrix:Matrix;
+	public var color:ColorTransform;
+	
+	public function canAddTriangles(numTriangles:Int):Bool
+	{
+		return true;
```
Comment (gamedevsam) [inline flixel/system/render/hardware/gl/FlxDrawTrianglesCommand.hx]: Shouldn't this return false since it's not supported?

File: flixel/system/render/hardware/gl/GLRenderHelper.hx [hunk 27]
```diff
@@ -0,0 +1,392 @@
+package flixel.system.render.hardware.gl;
```
Comment (gamedevsam) [inline flixel/system/render/hardware/gl/GLRenderHelper.hx]: Why is ```gl``` inside a ```hardware``` folder? Seems unnecessary.

File: flixel/graphics/FlxMaterial.hx (hunk 23 above)
Comment (Beeblerox) [inline flixel/graphics/FlxMaterial.hx]: I'm working on next iteration of renderer which uses this class. It will support multitexture rendering. Hope to finish it soon

File: flixel/system/render/hardware/FlxHardwareView.hx (hunk 24 above)
Comment (Beeblerox) [inline flixel/system/render/hardware/FlxHardwareView.hx]: it is used on flash target only

File: flixel/system/render/hardware/gl/FlxDrawTrianglesCommand.hx (hunk 26 above)
Comment (Beeblerox) [inline flixel/system/render/hardware/gl/FlxDrawTrianglesCommand.hx]: maybe :) 
I'll change it in the next pull request

File: flixel/system/render/hardware/gl/GLRenderHelper.hx (hunk 27 above)
Comment (Beeblerox) [inline flixel/system/render/hardware/gl/GLRenderHelper.hx]: do you think it's better to have 3 packages: `blit`, `tile` and `gl` on the same level?
Maybe it's good idea. Thanks!

**Comment by gamedevsam** [comment]
I forgot how to use git and pushed my fork of Beeblerox's RenderUpdate2 branch to the flixel repo, sorry everyone!

**Comment by Gama11** [comment]
Future discussion will happen in the new PR: #2068
